### PR TITLE
Refactor disequality processing in string solver

### PR DESCRIPTION
**Pull request by 4tXJ7f**
This commit refactors disequality processing in the core string solver.
It also adds statistics for the inferences and splits in those methods.
No semantic changes intended.

### CODE REVIEW

File: src/theory/strings/core_solver.cpp [hunk 1]
```diff
@@ -372,7 +372,7 @@ void CoreSolver::checkFlatForm(std::vector<Node>& eqc,
               // if lengths are the same, apply LengthEq
               std::vector<Node> lexp2;
               Node lcc = d_state.getLength(bc, lexp2);
-              if (d_state.areEqual(lcurr, lcc))
+              if (d_state.areEqual(lcurr, lcc) && true)
```
Comment (ajreynol) [inline src/theory/strings/core_solver.cpp:375]: I assume this was unintentional :)

File: src/theory/strings/core_solver.cpp [hunk 2]
```diff
@@ -1723,293 +1723,404 @@ CoreSolver::ProcessLoopResult CoreSolver::processLoop(NormalForm& nfi,
...
+  {
+    return;
+  }
+
+  std::vector<Node> nfi = nfni.d_nf;
+  std::vector<Node> nfj = nfnj.d_nf;
+
+  // See if one side is constant, if so, the disequality ni != nj is satisfied
+  // since ni does not contain nj or vice versa.
+  // This is only valid when isRev is false, since when isRev=true, the contents
```
Comment (ajreynol) [inline src/theory/strings/core_solver.cpp:1743]: Can delete this comment.

File: src/theory/strings/core_solver.cpp (hunk 1 above)
Comment (4tXJ7f) [reply src/theory/strings/core_solver.cpp:375]: Hahaha, yep :)

File: src/theory/strings/infer_info.h [hunk 3]
```diff
@@ -119,6 +119,35 @@ enum class Inference : uint32_t
...
+  DEQ_DISL_STRINGS_SPLIT,
+  // When x ++ x' ++ ... != y ++ y' ++ ... ^ len(x) = len(y), we apply the
+  // inference:
+  //   x = y v x != y
+  DEQ_STRINGS_EQ,
+  // When x ++ x' ++ ... != y ++ y' ++ ... and we do not know how the lengths
+  // of x and y compare, we apply the inference:
+  //   len(x) = len(y) v len(x) != len(y)
+  DEQ_LENS_EQ,
+  // When px ++ x ++ ... != py ^ len(px ++ x ++ ...) = len(py), we apply the
```
Comment (ajreynol) [inline src/theory/strings/infer_info.h:146]: This one does have an antecedant.

File: src/theory/strings/infer_info.h [hunk 4]
```diff
@@ -119,6 +119,35 @@ enum class Inference : uint32_t
...
+  DEQ_DISL_EMP_SPLIT,
+  // When x ++ x' ++ ... != "abc" ++ y' ++ ... ^ len(x) = 1, we apply the
+  // inference:
+  //   x = "a" v x != "a"
+  DEQ_DISL_FIRST_CHAR_EQ_SPLIT,
+  // When x ++ x' ++ ... != "abc" ++ y' ++ ... ^ len(x) != "", we apply the
+  // inference:
+  //   x = k1 ++ k2 ^ len(k1) = 1 ^ (k1 != "a" v x = "a" ++  k2)
+  DEQ_DISL_FIRST_CHAR_STRING_SPLIT,
+  // When x ++ x' ++ ... != y ++ y' ++ ... ^ len(x) != len(y), we apply the
```
Comment (ajreynol) [inline src/theory/strings/infer_info.h:134]: This one does have an antecedant.

File: src/theory/strings/infer_info.h [hunk 5]
```diff
@@ -119,6 +119,35 @@ enum class Inference : uint32_t
...
   FLOOP,
+  // When x ++ x' ++ ... != "abc" ++ y' ++ ... ^ len(x) != len(y), we apply the
+  // inference:
+  //   x = "" v x != ""
+  DEQ_DISL_EMP_SPLIT,
+  // When x ++ x' ++ ... != "abc" ++ y' ++ ... ^ len(x) = 1, we apply the
+  // inference:
+  //   x = "a" v x != "a"
+  DEQ_DISL_FIRST_CHAR_EQ_SPLIT,
+  // When x ++ x' ++ ... != "abc" ++ y' ++ ... ^ len(x) != "", we apply the
```
Comment (ajreynol) [inline src/theory/strings/infer_info.h:130]: This does have an antecendant.

File: src/theory/strings/core_solver.h [hunk 6]
```diff
@@ -314,16 +314,16 @@ class CoreSolver
 
   //--------------------------for checkNormalFormsDeq
   void processDeq(Node n1, Node n2);
-  int processReverseDeq(std::vector<Node>& nfi,
+  bool processReverseDeq(std::vector<Node>& nfi,
```
Comment (ajreynol) [inline src/theory/strings/core_solver.h:317]: Can you add a short description of these while they are fresh in memory?

**Pull request merged**